### PR TITLE
[SPARK-26412][PySpark][SQL] Allow Pandas UDF to take an iterator of pd.Series or an iterator of tuple of pd.Series

### PR DESCRIPTION
**Pull request by WeichenXu123**
## What changes were proposed in this pull request?

Allow Pandas UDF to take an iterator of pd.Series or an iterator of tuple of pd.Series.
Note the UDF input args will be always one iterator:
* if the udf take only column as input, the iterator's element will be pd.Series (corresponding to the column values batch)
* if the udf take multiple columns as inputs, the iterator's element will be a tuple composed of multiple `pd.Series`s, each one corresponding to the multiple columns as inputs (keep the same order). For example:
```
@pandas_udf("int", PandasUDFType.SCALAR_ITER)
def the_udf(iterator):
    for col1_batch, col2_batch in iterator:
        yield col1_batch + col2_batch

df.select(the_udf("col1", "col2"))
```
The udf above will add col1 and col2.

I haven't add unit tests, but manually tests show it works fine. So it is ready for first pass review.
We can test several typical cases:

```
from pyspark.sql import SparkSession
from pyspark.sql.functions import pandas_udf, PandasUDFType
from pyspark.sql.functions import udf
from pyspark.taskcontext import TaskContext

df = spark.createDataFrame([(1, 20), (3, 40)], ["a", "b"])

@pandas_udf("int", PandasUDFType.SCALAR_ITER)
def fi1(it):
    pid = TaskContext.get().partitionId()
    print("DBG: fi1: do init stuff, partitionId=" + str(pid))
    for batch in it:
        yield batch + 100
    print("DBG: fi1: do close stuff, partitionId=" + str(pid))

@pandas_udf("int", PandasUDFType.SCALAR_ITER)
def fi2(it):
    pid = TaskContext.get().partitionId()
    print("DBG: fi2: do init stuff, partitionId=" + str(pid))
    for batch in it:
        yield batch + 10000
    print("DBG: fi2: do close stuff, partitionId=" + str(pid))

@pandas_udf("int", PandasUDFType.SCALAR_ITER)
def fi3(it):
    pid = TaskContext.get().partitionId()
    print("DBG: fi3: do init stuff, partitionId=" + str(pid))
    for x, y in it:
        yield x + y * 10 + 100000
    print("DBG: fi3: do close stuff, partitionId=" + str(pid))


@pandas_udf("int", PandasUDFType.SCALAR)
def fp1(x):
    return x + 1000

@udf("int")
def fu1(x):
    return x + 10

# test select "pandas iter udf/pandas udf/sql udf" expressions at the same time.
# Note this case the `fi1("a"), fi2("b"), fi3("a", "b")` will generate only one plan,
# and `fu1("a")`, `fp1("a")` will generate another two separate plans.
df.select(fi1("a"), fi2("b"), fi3("a", "b"), fu1("a"), fp1("a")).show()

# test chain two pandas iter udf together
# Note this case `fi2(fi1("a"))` will generate only one plan
# Also note the init stuff/close stuff call order will be like:
# (debug output following)
#     DBG: fi2: do init stuff, partitionId=0
#     DBG: fi1: do init stuff, partitionId=0
#     DBG: fi1: do close stuff, partitionId=0
#     DBG: fi2: do close stuff, partitionId=0
df.select(fi2(fi1("a"))).show()

# test more complex chain
# Note this case `fi1("a"), fi2("a")` will generate one plan,
# and `fi3(fi1_output, fi2_output)` will generate another plan
df.select(fi3(fi1("a"), fi2("a"))).show()
```

## How was this patch tested?

To be added.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by WeichenXu123** [comment]
I split the "add per batch flush" changing into a new PR https://github.com/apache/spark/pull/24734

File: python/pyspark/worker.py [hunk 1]
```diff
@@ -86,21 +86,29 @@ def wrap_udf(f, return_type):
         return lambda *a: f(*a)
 
 
-def wrap_scalar_pandas_udf(f, return_type):
+def verify_scalar_pandas_udf_result_length(result, length):
```
Comment (mengxr) [inline python/pyspark/worker.py]: This method is not called outside `wrap_scalar_pandas_udf`. Could you define it inside and rename it to `verify_result_length`?

File: python/pyspark/worker.py [hunk 2]
```diff
@@ -86,21 +86,29 @@ def wrap_udf(f, return_type):
...
-            raise RuntimeError("Result vector from pandas_udf was not the required length: "
-                               "expected %d, got %d" % (len(a[0]), len(result)))
         return result
 
-    return lambda *a: (verify_result_length(*a), arrow_return_type)
+    if eval_type == PythonEvalType.SQL_SCALAR_PANDAS_UDF:
+        return lambda *a: (verify_scalar_pandas_udf_result_length(
+            verify_result_type(f(*a)), len(a[0])), arrow_return_type)
+    else:
+        return lambda *iterator: map(lambda res: (res, arrow_return_type),
```
Comment (mengxr) [inline python/pyspark/worker.py]: Put an inline comment to mention the result length verification is done at the end of a partition.

File: python/pyspark/worker.py [hunk 3]
```diff
@@ -255,13 +269,54 @@ def read_udfs(pickleSer, infile, eval_type):
...
     num_udfs = read_int(infile)
+
+    if eval_type == PythonEvalType.SQL_SCALAR_PANDAS_ITER_UDF:
+        assert num_udfs == 1, "One SQL_SCALAR_PANDAS_ITER_UDF expected here."
+
+        arg_offsets, udf = read_single_udf(
+            pickleSer, infile, eval_type, runner_conf, udf_index=i)
+
+        def func(_, iterator):
+            num_input_rows = [0]
```
Comment (mengxr) [inline python/pyspark/worker.py]: define `num_input_rows` outside func and then you shouldn't need to wrap the value inside array

File: python/pyspark/worker.py [hunk 4]
```diff
@@ -255,13 +269,54 @@ def read_udfs(pickleSer, infile, eval_type):
...
+                    return tuple(udf_args)
+
+            iterator = map(map_batch, iterator)
+            result_iter = udf(iterator)
+
+            num_output_rows = 0
+            for result_batch, result_type in result_iter:
+                num_output_rows += len(result_batch)
+                assert num_output_rows <= num_input_rows[0], \
+                    "Pandas iterator UDF generate more rows then read rows."
```
Comment (mengxr) [inline python/pyspark/worker.py]: `Pandas SCALAR_ITER UDF outputted more rows than input rows.`

File: python/pyspark/worker.py [hunk 5]
```diff
@@ -255,13 +269,54 @@ def read_udfs(pickleSer, infile, eval_type):
...
+                yield (result_batch, result_type)
+            try:
+                iterator.__next__()
+                raise Exception("SQL_SCALAR_PANDAS_ITER_UDF should exhaust the input iterator.")
```
Comment (mengxr) [inline python/pyspark/worker.py]:
~~~python
try:
  iterator.__next__()
except StopIteration:
  pass
else:
  raise RuntimeError("SQL_SCALAR_PANDAS_ITER_UDF should exhaust the input iterator.")
~~~

File: python/pyspark/worker.py [hunk 6]
```diff
@@ -255,13 +269,54 @@ def read_udfs(pickleSer, infile, eval_type):
...
+            except StopIteration:
+                pass
+
+            if num_output_rows != num_input_rows[0]:
+                raise Exception("The number of output rows of the pandas iterator UDF should be "
```
Comment (mengxr) [inline python/pyspark/worker.py]: Use `RuntimeError` and put number of input/output rows in the error message.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala [hunk 7]
```diff
@@ -111,19 +111,27 @@ object ExtractPythonUDFs extends Rule[LogicalPlan] with PredicateHelper {
...
-    // shouldn't change later.
-    // Used to check if subsequent UDFs are of the same type as the first UDF. (since we can only
+    // If fisrt UDF is SQL_SCALAR_PANDAS_ITER_UDF, then only return this UDF,
+    // otherwise check if subsequent UDFs are of the same type as the first UDF. (since we can only
     // extract UDFs of the same eval type)
-    var evalTypeChecker: Option[EvalTypeChecker] = None
+
+    var firstVisitedUDFEvalType: Option[Int] = None
+
+    def checkEvalType(evalType: Int): Boolean = {
```
Comment (mengxr) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala]: rename it to `canChainUDF`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala [hunk 8]
```diff
@@ -175,16 +183,20 @@ object ExtractPythonUDFs extends Rule[LogicalPlan] with PredicateHelper {
...
 
-          val evaluation = validUdfs.partition(
-            _.evalType == PythonEvalType.SQL_SCALAR_PANDAS_UDF
-          ) match {
-            case (vectorizedUdfs, plainUdfs) if plainUdfs.isEmpty =>
-              ArrowEvalPython(vectorizedUdfs, resultAttrs, child)
-            case (vectorizedUdfs, plainUdfs) if vectorizedUdfs.isEmpty =>
-              BatchEvalPython(plainUdfs, resultAttrs, child)
-            case _ =>
+          val evalType = validUdfs.head.evalType
```
Comment (mengxr) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala]:
~~~scala
val evalTypes = validUdfs.map(_.evalType).toSet
if (evalTypes.size != 1) {
  // .. include `evalTypes` in the error message
}
~~~

**Review by mengxr — CHANGES_REQUESTED** [review]
The overall implementation looks good to me. Please address comments, add tests, and document the new UDF type.

File: python/pyspark/worker.py (hunk 3 above)
Comment (WeichenXu123) [reply python/pyspark/worker.py]: define `num_input_rows` outside func doesn't work.
The closure variable can only be referenced from inner function, but cannot be assigned new value.
Another working approach is we can add a declaration `nonlocal num_input_rows`, but python2 do not support this.

**Comment by WeichenXu123** [comment]
@meng Done. Leave doc which will update soon.

**Comment by icexelloss** [comment]
Pretty cool stuff! Just want to leave a few thoughts:

The proposed API doesn't seem to point out the notion of "partition", which I think it important here. In other words, just by looking at the API, it's not clear to me that the UDF is going to be invoked once per partition.

Another thought is that exposing the notion of partition isn't really something common in UDFs. I wonder why do we prefer this over sth like mapPartitions?

**Comment by mengxr** [comment]
@icexelloss I'm not sure if we should claim "execute once per partition" semantic even it is currently implemented this way. UDFs are on the logical side while per-partition execution is on the physical side. It should be the same result if we split the partition into two can call the new UDF twice. I'm assuming the map function is deterministic here.

**Review by mengxr — CHANGES_REQUESTED** [review]
Some minor comments.

File: python/pyspark/sql/tests/test_pandas_udf_scalar.py [hunk 9]
```diff
@@ -136,84 +150,115 @@ def test_register_nondeterministic_vectorized_udf_basic(self):
         [row] = self.spark.sql("SELECT randomPandasUDF(1)").collect()
         self.assertEqual(row[0], 7)
 
+        def random_iter_udf(it):
+            for i in it:
+                yield random.randint(6, 6) + i
```
[the remark this answers is not in the document]
Comment (WeichenXu123) [reply python/pyspark/sql/tests/test_pandas_udf_scalar.py]: The master test code is `random.randint(6, 6)`, I guess they want the test to keep deterministic. The "random.randint" is just a hint for the udf is random ?

**Comment by mengxr** [comment]
I added two tests that verifies init and close with the new UDF. LGTM pending Jenkins.

**Comment by mengxr** [comment]
I created https://issues.apache.org/jira/browse/SPARK-28056 for the follow-up work on documentation.

**Comment by mengxr** [comment]
test this please

**Comment by mengxr** [comment]
jenkins test this please

**Comment by jiangxb1987** [comment]
retest this please

**Comment by mengxr** [comment]
Merged into master. Thanks!

**Pull request closed without merging**